x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,728 @@
+  author: RuleAuthorArray,
+  buildingBlockType: BuildingBlockType.optional(),
+  description: RuleDescription,
+  namespace: AlertsIndexNamespace.optional(),
+  note: InvestigationGuide.optional(),
+  falsePositives: RuleFalsePositiveArray,
+  from: RuleIntervalFrom,
+  ruleId: RuleSignatureId,
+  investigationFields: InvestigationFieldsCombined.optional(),
+  // Deprecated, migrated to "prebuilt"
+  immutable: IsRuleImmutable,
+  prebuilt: IsRulePrebuilt,
+  customized: IsPrebuiltRuleCustomized,
+  license: RuleLicense.optional(),
+  outputIndex: AlertsIndex,
+  timelineId: TimelineTemplateId.optional(),
+  timelineTitle: TimelineTemplateTitle.optional(),
+  meta: RuleMetadata.optional(),
+  maxSignals: MaxSignals,
+  riskScore: RiskScore,
+  riskScoreMapping: RiskScoreMapping,
+  ruleNameOverride: RuleNameOverride.optional(),
+  severity: Severity,
+  severityMapping: SeverityMapping,
+  timestampOverride: TimestampOverride.optional(),
+  timestampOverrideFallbackDisabled: TimestampOverrideFallbackDisabled.optional(),
+  threat: ThreatArray,
+  to: RuleIntervalTo,
+  references: RuleReferenceArray,
+  version: RuleVersion,
+  exceptionsList: RuleExceptionList.array(),
+  relatedIntegrations: RelatedIntegrationArray.optional(),
+  requiredFields: RequiredFieldArray.optional(),
+  setup: SetupGuide.optional(),
+});
+```
+
+```yaml
+// common_attributes.schema.yaml
+[...]
+    IsRulePrebuilt:
+      type: boolean
+      description: Determines whether the rule is an Elastic Prebuilt Rule that can receive upstream updates via Fleet.
+
+    IsPrebuiltRuleCustomized:
+      type: boolean
+      description: Determines whether the prebuilt rule has been customized by the user (i.e. any of its fields have been modified and diverged from the base value).
+[...]
+```
+
+## Mapping changes
+
+**Alert (rule objects) mapping**
+
+No changes will be needed for the [mapping of rule saved objects](https://github.com/elastic/kibana/blob/main/x-pack/plugins/alerting/common/saved_objects/rules/mappings.ts) (of type `alert`), since the new fields introduced will be part of the `params` field, which is a `flattened` field.
+
+**Security Rules (prebuilt rule assets) mapping**
+
+No changes will be needed either for the `security-rule` [mapping](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/rule_assets/prebuilt_rule_assets_type.ts), our prebuilt rule assets. We currently have mappings for the `rule_id` and `version` fields, since we perform aggregations and filtering on those in the rule install and upgrade endpoints. No additional fields needs to be mapped in this phase.
+
+## Plan for carrying out migrations of rule SOs
+
+### Migration strategy: perform migration on all write operations of the rules
+
+In order to migrate our users rule's data to the new schema, we should perform the migration of rules whenever any endpoint operation that writes/updates a rule saved object.
```

Comment:
> Let's provide readers of this RFC with context and explain why we have to migrate rules in our endpoints instead of using the migration mechanism.

Added this explanation in the `### Context` section above. I didn't write any migrations with the historical migration mechanism, so I have limited understanding of how it was carried out, but I think I described it more or less correctly. Can you take a look and tell me if anything I described there is wrong?

> Also, let's distinguish right from the beginning the two types of this migration:

Yes, makes sense. I'll rewrite this section with a clear split between them, and explaining them separately.